File: matlab/toolbox_signal/perform_wavortho_transf.m
function f = perform_wavortho_transf(f,Jmin,dir,options)

% perform_wavortho_transf - compute orthogonal wavelet transform
%
%   fw = perform_wavortho_transf(f,Jmin,dir,options);
%
%   You can give the filter in options.h.
%
%   Works in arbitrary dimension.
%
%   Copyright (c) 2009 Gabriel Peyre

options.null = 0;
h = getoptions(options,'h', compute_wavelet_filter('Daubechies',4) );
g = [0 h(length(h):-1:2)] .* (-1).^(1:length(h));

n = size(f,1); 
Jmax = log2(n)-1; 

if dir==1
    %%% FORWARD %%%
    for j=Jmax:-1:Jmin
        sel = 1:2^(j+1);
        a = subselect(f,sel);
        for d=1:nb_dims(f)
            a = cat(d, subsampling(cconv(a,h,d),d), subsampling(cconv(a,g,d),d) );
        end
        f = subassign(f,sel,a);
    end
else
    %%% FORWARD %%%
    for j=Jmin:Jmax
        sel = 1:2^(j+1);
        a = subselect(f,sel);
        for d=1:nb_dims(f)
            w = subselectdim(a,2^j+1:2^(j+1),d);
            a = subselectdim(a,1:2^j,d);
            a = cconv(upsampling(a,d),reverse(h),d) + cconv(upsampling(w,d),reverse(g),d);
        end
        f = subassign(f,sel,a);
    end    
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function f = subselect(f,sel)
switch nb_dims(f)
    case 1
        f = f(sel);
    case 2
        f = f(sel,sel);
    case 3
        f = f(sel,sel,sel);
    case 4
        f = f(sel,sel,sel,sel);
    case 5
        f = f(sel,sel,sel,sel,sel);
    case 6
        f = f(sel,sel,sel,sel,sel,sel);
    case 7
        f = f(sel,sel,sel,sel,sel,sel,sel);
    case 8
        f = f(sel,sel,sel,sel,sel,sel,sel,sel);
    otherwise
        error('Not implemented');
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function f = subselectdim(f,sel,d)
switch d
    case 1
        f = f(sel,:,:,:,:,:,:,:);
    case 2
        f = f(:,sel,:,:,:,:,:,:);
    case 3
        f = f(:,:,sel,:,:,:,:,:);
    case 4
        f = f(:,:,:,sel,:,:,:,:);
    case 5
        f = f(:,:,:,:,sel,:,:,:);
    case 6
        f = f(:,:,:,:,:,sel,:,:);
    case 7
        f = f(:,:,:,:,:,:,sel,:);
    case 8
        f = f(:,:,:,:,:,:,:,sel);
    otherwise
        error('Not implemented');
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function f = subassign(f,sel,g)
switch nb_dims(f)
    case 1
        f(sel) = g;
    case 2
        f(sel,sel) = g;
    case 3
        f(sel,sel,sel) = g;
    case 4
        f(sel,sel,sel,sel) = g;
    case 5
        f(sel,sel,sel,sel,sel) = g;
    case 6
        f(sel,sel,sel,sel,sel,sel) = g;
    case 7
        f(sel,sel,sel,sel,sel,sel,sel) = g;
    case 8
        f(sel,sel,sel,sel,sel,sel,sel,sel) = g;
    otherwise
        error('Not implemented');
end

File: r/nt_toolbox/toolbox_signal/perform_wavortho_transf.R




perform_wavortho_transf <- function(f, Jmin, dir, h){
  ####
  # perform_wavortho_transf - compute orthogonal wavelet transform
  
  # fw = perform_wavortho_transf(f,Jmin,dir,options);
  
  # You can give the filter in options.h.
  
  # Works in 2D only.
  
  # Copyright (c) 2014 Gabriel Peyre
  ####
  
  
  n <- dim(f)[2]
  Jmax <- as.integer(log(n, 2) - 1)
  # compute g filter
  u <- (- rep(1, length(h) - 1)) ** (1:(length(h)-1))
  # alternate +1/-1
  g <- c(0, rev(h[2:length(h)]) * u)
  
  if (dir == 1){
    ### FORWARD ###
    fW <- as.matrix(f)
    for (j in Jmax:Jmin){
      A <- fW[1:2**(j+1), 1:2**(j+1)]
      for (d in (1:2)){
        Coarse <- subsampling(cconv(A, h, d), d)
        Detail <- subsampling(cconv(A, g, d), d)
        if (d==1){
          A <- rbind(Coarse, Detail)
        }
        else{
          A <- cbind(Coarse, Detail)
        }
      }
      fW[1:2 ** (j + 1), 1:2 ** (j + 1)] <- A
    }
    return(fW)
  }
  else{
    ### BACKWARD ###
    fW <- f
    f1 <- fW
    for (j in Jmin:Jmax){
      A <- f1[1:2 ** (j + 1), 1:2 ** (j + 1)] ###
      for (d in (1:2)){
        if (d==1){
          Coarse <- A[1:2**j, , drop=F]
          Detail <- A[(2**j + 1): 2**(j + 1), , drop=F]
        }
        else{
          Coarse <- A[ , 1:2 ** j, drop=F]
          Detail <- A[ , (2 ** j + 1):2 ** (j + 1), drop=F]
        }
        Coarse <- cconv(upsampling(Coarse, d), reverse(h), d)
        Detail <- cconv(upsampling(Detail, d), reverse(g), d)
        A <- Coarse + Detail
      }
      f1[1:2 ** (j + 1), 1:2 ** (j + 1)] <- A
    }
    return(f1)
  }
}

